src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java

```diff
@@ -693,6 +697,13 @@ private static float height(List<RowGroup> rg, boolean useLastUnitSize) {
     private BlockContext addRows(List<RowGroup> head, PageImpl p, BlockContext blockContext) {
         int i = head.size();
         for (RowGroup rg : head) {
+
+            // At this point we expect that the attribute keep="page" is set, which means that each
+            // block should not be spanning multiple pages. This is required so we don't print just a part
+            // of the block when the display-when attribute is set to false.
+            // We also expect the display-when value never to be true. It could be missing or an expression
+            // which will remove a block at the beginning of a page. (! $starts-at-top-of-page)
+            // These restrictions are added in the OBFL Parser and will lead to exceptions being thrown.
```

Comment:
> At this point we expect that the attribute keep="page" is set

We expect keep="page" is set _when the condition evaluates to false_.

> We also expect the display-when value never to be true.

We expect display-when is either set to "true" (or missing, which is the same) or "(! $starts-at-top-of-page)". Other values are not permitted.

Above assumption deserves some more explanation for a good understanding: we need it because we evaluate display-when _for each RowGroup_ while normally it should be evaluated _only _once per block_. This is fine in the two mentioned cases:
- display-when="true": trivial.
- display-when="(! $starts-at-top-of-page)":
  - if this evaluates to true, it means the page must already contain a RowImpl(*), so regardless of whether the current RowGroup belongs to the same block or a new block, it must be rendered.
  - if it evaluates to false, it means the page does not already contain a RowImpl, so we know it must be the start of a new page and block (because keep="page" in this case).

(*) Refer to the line below where you set `.topOfPage(false)`. The line should however be moved inside the for loop because it could be that all of RowGroups are skipped, in which case the value should remain the same.